samples/s03-singleAABB/hostCode.cpp

```diff
@@ -0,0 +1,267 @@
+
+#define LOG(message)                                            \
+  std::cout << GPRT_TERMINAL_BLUE;                               \
+  std::cout << "#gprt.sample(main): " << message << std::endl;   \
+  std::cout << GPRT_TERMINAL_DEFAULT;
+#define LOG_OK(message)                                         \
+  std::cout << GPRT_TERMINAL_LIGHT_BLUE;                         \
+  std::cout << "#gprt.sample(main): " << message << std::endl;   \
+  std::cout << GPRT_TERMINAL_DEFAULT;
+
+extern GPRTProgram s03_deviceCode;
+
+// The extents of our bounding box
+float3 aabbPositions[2] =
+  {
+    {-1.0f, -1.0f, -1.0f},
+    { 1.0f,  1.0f,  1.0f}
+  };
+
+// initial image resolution
+const int2 fbSize = {1400, 460};
+
+// final image output
+const char *outFileName = "s03-singleAABB.png";
+
+// Initial camera parameters
+float3 lookFrom = {3.5f,3.5f,3.5f};
+float3 lookAt = {0.f,0.f,0.f};
+float3 lookUp = {0.f,1.f,0.f};
+float cosFovy = 0.66f;
+
+#include <iostream>
+int main(int ac, char **av)
+{
+  // In this example, we will create an axis aligned bounding box (AABB). These
+  // are more general than triangle primitives, and can be used for custom 
+  // primitive types.
+  LOG("gprt example '" << av[0] << "' starting up");
+
+  // create a context on the first device:
+  gprtRequestWindow(fbSize.x, fbSize.y, "S03 Single AABB");
+  GPRTContext context = gprtContextCreate();
+  GPRTModule module = gprtModuleCreate(context,s03_deviceCode);
+
+  // ##################################################################
+  // set up all the GPU kernels we want to run
+  // ##################################################################
+
+  // -------------------------------------------------------
+  // declare geometry type
+  // -------------------------------------------------------
+  GPRTVarDecl aabbGeomVars[] = {
+    { /* sentinel to mark end of list */ }
+  };
+  GPRTGeomType aabbGeomType
+    = gprtGeomTypeCreate(context,
+                        GPRT_AABBS, // <- This is new!
+                        sizeof(AABBGeomData),
+                        aabbGeomVars,-1);
+  gprtGeomTypeSetClosestHitProg(aabbGeomType,0,
+                           module,"AABBClosestHit");
+  gprtGeomTypeSetIntersectionProg(aabbGeomType,0,
+                           module,"AABBIntersection");
+
+  // -------------------------------------------------------
+  // set up miss
+  // -------------------------------------------------------
+  GPRTVarDecl missVars[]
+    = {
+    { "color0", GPRT_FLOAT3, GPRT_OFFSETOF(MissProgData,color0)},
+    { "color1", GPRT_FLOAT3, GPRT_OFFSETOF(MissProgData,color1)},
+    { /* sentinel to mark end of list */ }
+  };
+  GPRTMiss miss
+    = gprtMissCreate(context,module,"miss",sizeof(MissProgData),
+                        missVars,-1);
+
+  // -------------------------------------------------------
+  // set up ray gen program
+  // -------------------------------------------------------
+  GPRTVarDecl rayGenVars[] = {
```

Comment:
Starting to see where some "boilerplate" structs that we keep in GPRT might be useful here...

Reply:
But for the tutorials (at least the starting ones) it's probably instructive to do all of this explicitly.

Reply:
yeah, for now, lets keep things clear and explicit to avoid confusion.